grails-app/controllers/com/unifina/controller/api/KeyApiController.groovy

```diff
@@ -126,22 +127,53 @@ class KeyApiController {
 	}
 
 	@StreamrApi(authenticationLevel = AuthLevel.USER)
-	def update() {
-		Key k = Key.get(params.id)
+	def updateUserKey() {
+		Key k = Key.get(params.keyId)
 		if (k == null) {
-			throw new NotFoundException(k.toString(), params.id)
+			throw new NotFoundException(k.toString(), params.keyId)
 		}
 		Map json = new JsonSlurper().parseText((String) request.JSON)
 		if (json.name != null && json.name.trim() != "") {
-			useResource(params.resourceClass, params.resourceId) { res ->
+			useResource(SecUser, params.keyId) { res ->
 				k.name = json.name.trim()
 				k.save(flush: true, failOnError: true)
 			}
 		}
 		response.status = 200
-		render([
-			id: k.id,
-			name: k.name,
-		] as JSON)
+		render(k.toMap() as JSON)
+	}
+
+	@StreamrApi(authenticationLevel = AuthLevel.USER)
+	def updateStreamKey() {
+		Key k = Key.get(params.keyId)
+		if (k == null) {
+			throw new NotFoundException(k.toString(), params.keyId)
+		}
+		Map json = new JsonSlurper().parseText((String) request.JSON)
+		String permission = request.JSON?.permission
+		if (permission) {
+			if (permission != "read" && permission != "write") {
+				throw new ApiException(400, "INVALID_PARAMETER", "permission field in json should be 'read' or 'write'.")
+			}
+		}
+		if (json.name != null && json.name.trim() != "") {
+			useResource(Stream, params.streamId) { res ->
+				k.name = json.name.trim()
+				if (permission) {
+					Permission.Operation operation = Permission.Operation.fromString(permission)
+					k.getPermissions().collect().each { p ->
+						k.removeFromPermissions(p)
```

Comment:
This works but causes unnecessary churn in the database. For example, if the user just updates the key name, the existing permission rows get deleted and new ones created, for no reason. Would be better to first see if any changes need to be done, and then do the minimal changes. Something like (pseudocode):
```
if (operation == Permission.Operation.READ) {
    grant read if it doesn't exist already
    delete write if it exists
} else if (operation == Permission.Operation.WRITE) {
    grant read if it doesn't exist already
    grant write if it doesn't exist already
}
```